onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -49,10 +49,25 @@ class Memcpy final : public OpKernel {
       auto X_dtype = X->DataType();
       Y->SetType(X_dtype);
       AllocatorPtr alloc;
-      auto status = ctx->GetTempSpaceAllocator(&alloc);
-      if (!status.IsOK()) {
-        return Status(common::ONNXRUNTIME, common::FAIL,
-                      "Memcpy cuda: unable to get an allocator.");
+
+      // If we are copying contents to CUDA, the allocator to use
+      // to allocate the buffers of the new tensors in the sequence
+      // can be temp space allocator associated with the CUDA EP
+      if (Node().OpType() == "MemcpyFromHost") {
+        auto status = ctx->GetTempSpaceAllocator(&alloc);
+        if (!status.IsOK()) {
+          return Status(common::ONNXRUNTIME, common::FAIL,
+                        "Memcpy cuda: unable to get an allocator.");
+        }
+      } else {
+        // If we are copying contents to CPU (op type is "MemcpyToHost"),
+        // the allocator to use to allocate the buffers of the new tensors
+        // in the sequence will be the allocator from the CPU EP
+        auto status = ctx->GetCPUAllocator(&alloc);
+        if (!status.IsOK()) {
+          return Status(common::ONNXRUNTIME, common::FAIL,
+                        "Memcpy cuda: unable to get the CPU allocator.");
+        }
       }
       auto X_size = X->Size();
```

Comment:
Would it be cleaner to have the data transfer manager deal with this like it assumably does for the other types?

Reply:
Not quite following. Doesn't the data transfer manager do the actual copy here ?

Reply:
Sorry - CodeFlow seems to have moved the comment. 

I was wondering if data transfer manager should provide a facility for copying sequence tensors so that all the logic required is in one place and the usage is similar to copying the other tensor types. 

I guess we can consider that later if we have to add this logic to another EP.

Reply:
Yeah, I prefer to keep the amount of change minimal so as to just unblock the release if that is okay.